Duplicate requests due to speculative parsing with active service worker

youennf opened the pull request:
#### c68a38f2d6ebefb7a1d63d53ee6e5456327dd34a
<pre>
Duplicate requests due to speculative parsing with active service worker
<a href="https://bugs.webkit.org/show_bug.cgi?id=250454">https://bugs.webkit.org/show_bug.cgi?id=250454</a>
<a href="https://rdar.apple.com/104377727">rdar://104377727</a>

Reviewed by Chris Dumez.

In case a service worker is located in the same process as a controlled document, they both share the same memory cache.
Before the patch, in case of preloads triggered byt he speculative parser and the preloads are not cacheable, the following would happen:
- Trigger a preload, which gets registered in the memory cache.
- Let the service worker handle the preload, which triggers a fetch of the same resource.
- This fetch will remove the preload resource from the memory cache and do the actual network load.
- The real load of the HTML resource kicks in and tries to load the resource from the memory cache.
- Given the resource is not a preload and is not reusable according the HTTP headers, a new load is triggered.

This makes it so that two loads are happening for the same resource (one for the preload and one for the actual load).
To resolve the issue, we do not add the service worker fetch resource in the memory cache in the following case:
- There is an existing cached resource, which is a preload, and not a service worker request.
- The new request is a service worker request.

This ensures that we keep the preload cached resource in the memory cache which allows reuse when the actual load happens.

* LayoutTests/http/wpt/service-workers/resources/serviceworker-and-preloads-iframe.html: Added.
* LayoutTests/http/wpt/service-workers/resources/serviceworker-preload-img.png: Added.
* LayoutTests/http/wpt/service-workers/resources/serviceworker-preload-img.png.headers: Added.
* LayoutTests/http/wpt/service-workers/resources/serviceworker-preload-script.js: Added.
* LayoutTests/http/wpt/service-workers/resources/serviceworker-preload-script.js.headers: Added.
* LayoutTests/http/wpt/service-workers/serviceworker-and-preloads-expected.txt: Added.
* LayoutTests/http/wpt/service-workers/serviceworker-and-preloads-worker.js: Added.
* LayoutTests/http/wpt/service-workers/serviceworker-and-preloads.html: Added.
* Source/WebCore/loader/cache/CachedResourceLoader.cpp:
(WebCore::CachedResourceLoader::updateCachedResourceWithCurrentRequest):
(WebCore::computeShouldAddToMemoryCache):
(WebCore::CachedResourceLoader::requestResource):
(WebCore::CachedResourceLoader::loadResource):
* Source/WebCore/loader/cache/CachedResourceLoader.h:

Canonical link: <a href="https://commits.webkit.org/272024@main">https://commits.webkit.org/272024@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/09c7f96373d5575900f1e31d58d1c2060f9676a0

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/30266 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/8937 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/31933 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/32774 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/27386 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/30957 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/11135 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/6174 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/27378 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/30573 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/7519 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/27098 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/6428 "Passed tests") | 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/6589 "Too many flaky failures: imported/w3c/web-platform-tests/css/css-cascade/scope-hover.html, imported/w3c/web-platform-tests/css/css-view-transitions/massive-element-below-viewport-partially-onscreen-new.html, imported/w3c/web-platform-tests/css/css-view-transitions/massive-element-below-viewport-partially-onscreen-old.html, imported/w3c/web-platform-tests/css/selectors/invalidation/user-valid-user-invalid.html, imported/w3c/web-platform-tests/css/selectors/user-invalid.html, imported/w3c/web-platform-tests/css/selectors/valid-invalid-form-fieldset.html, imported/w3c/web-platform-tests/feature-policy/feature-policy-frame-policy-allowed-for-all.https.sub.html, imported/w3c/web-platform-tests/html/semantics/disabled-elements/disabled-event-dispatch.tentative.html, imported/w3c/web-platform-tests/html/semantics/invokers/invoketarget-button-event-dispatch.tentative.html, imported/w3c/web-platform-tests/html/semantics/selectors/pseudo-classes/readwrite-readonly.html, imported/w3c/web-platform-tests/screen-orientation/lock-unlock-check.html (failure)") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/27033 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/34114 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/27603 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/27533 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/32780 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/6539 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/4722 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/30578 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/8287 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/26788 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7198 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/7293 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/7065 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/loader/cache/CachedResourceLoader.cpp:
```diff
@@ -1248,17 +1255,17 @@ CachedResourceHandle<CachedResource> CachedResourceLoader::revalidateResource(Ca
     auto resource = createResource(type, WTFMove(request), sessionID, &cookieJar, settings);
 
-    if (resource->allowsCaching())
+    if (resource->allowsCaching() && shouldAddToMemoryCache == ShouldAddToMemoryCache::Yes)
```
Given that we still respect `resource->allowsCaching()` even if `ShouldAddToMemoryCache::Yes`, the current naming is a little bit confusing. Maybe the flag should be `MayAddToMemoryCache`

youennf committed:
Duplicate requests due to speculative parsing with active service worker
https://bugs.webkit.org/show_bug.cgi?id=250454
rdar://104377727

Reviewed by Chris Dumez.

In case a service worker is located in the same process as a controlled document, they both share the same memory cache.
Before the patch, in case of preloads triggered byt he speculative parser and the preloads are not cacheable, the following would happen:
- Trigger a preload, which gets registered in the memory cache.
- Let the service worker handle the preload, which triggers a fetch of the same resource.
- This fetch will remove the preload resource from the memory cache and do the actual network load.
- The real load of the HTML resource kicks in and tries to load the resource from the memory cache.
- Given the resource is not a preload and is not reusable according the HTTP headers, a new load is triggered.

This makes it so that two loads are happening for the same resource (one for the preload and one for the actual load).
To resolve the issue, we do not add the service worker fetch resource in the memory cache in the following case:
- There is an existing cached resource, which is a preload, and not a service worker request.
- The new request is a service worker request.

This ensures that we keep the preload cached resource in the memory cache which allows reuse when the actual load happens.

* LayoutTests/http/wpt/service-workers/resources/serviceworker-and-preloads-iframe.html: Added.
* LayoutTests/http/wpt/service-workers/resources/serviceworker-preload-img.png: Added.
* LayoutTests/http/wpt/service-workers/resources/serviceworker-preload-img.png.headers: Added.
* LayoutTests/http/wpt/service-workers/resources/serviceworker-preload-script.js: Added.
* LayoutTests/http/wpt/service-workers/resources/serviceworker-preload-script.js.headers: Added.
* LayoutTests/http/wpt/service-workers/serviceworker-and-preloads-expected.txt: Added.
* LayoutTests/http/wpt/service-workers/serviceworker-and-preloads-worker.js: Added.
* LayoutTests/http/wpt/service-workers/serviceworker-and-preloads.html: Added.
* Source/WebCore/loader/cache/CachedResourceLoader.cpp:
(WebCore::CachedResourceLoader::updateCachedResourceWithCurrentRequest):
(WebCore::computeShouldAddToMemoryCache):
(WebCore::CachedResourceLoader::requestResource):
(WebCore::CachedResourceLoader::loadResource):
* Source/WebCore/loader/cache/CachedResourceLoader.h:

Canonical link: https://commits.webkit.org/272024@main
